File: go/test/comment_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/korean-json-sdk/go"
	"github.com/voxgig-sdk/korean-json-sdk/go/core"

	vs "github.com/voxgig-sdk/korean-json-sdk/go/utility/struct"
)

func TestCommentEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Comment(nil)
		if ent == nil {
			t.Fatal("expected non-nil CommentEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := commentBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "comment." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set KOREANJSON_TEST_COMMENT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		commentRef01Ent := client.Comment(nil)
		commentRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "comment"}, setup.data), "comment_ref01"))

		commentRef01DataResult, err := commentRef01Ent.Create(commentRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		commentRef01Data = core.ToMapAny(commentRef01DataResult)
		if commentRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if commentRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		commentRef01Match := map[string]any{}

		commentRef01ListResult, err := commentRef01Ent.List(commentRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		commentRef01List, commentRef01ListOk := commentRef01ListResult.([]any)
		if !commentRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", commentRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(commentRef01List), map[string]any{"id": commentRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		commentRef01DataUp0Up := map[string]any{
			"id": commentRef01Data["id"],
		}

		commentRef01MarkdefUp0Name := "content"
		commentRef01MarkdefUp0Value := fmt.Sprintf("Mark01-comment_ref01_%d", setup.now)
		commentRef01DataUp0Up[commentRef01MarkdefUp0Name] = commentRef01MarkdefUp0Value

		commentRef01ResdataUp0Result, err := commentRef01Ent.Update(commentRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		commentRef01ResdataUp0 := core.ToMapAny(commentRef01ResdataUp0Result)
		if commentRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if commentRef01ResdataUp0["id"] != commentRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if commentRef01ResdataUp0[commentRef01MarkdefUp0Name] != commentRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", commentRef01MarkdefUp0Name, commentRef01ResdataUp0[commentRef01MarkdefUp0Name])
		}

		// LOAD
		commentRef01MatchDt0 := map[string]any{
			"id": commentRef01Data["id"],
		}
		commentRef01DataDt0Loaded, err := commentRef01Ent.Load(commentRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		commentRef01DataDt0LoadResult := core.ToMapAny(commentRef01DataDt0Loaded)
		if commentRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if commentRef01DataDt0LoadResult["id"] != commentRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		commentRef01MatchRm0 := map[string]any{
			"id": commentRef01Data["id"],
		}
		_, err = commentRef01Ent.Remove(commentRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		commentRef01MatchRt0 := map[string]any{}

		commentRef01ListRt0Result, err := commentRef01Ent.List(commentRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		commentRef01ListRt0, commentRef01ListRt0Ok := commentRef01ListRt0Result.([]any)
		if !commentRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", commentRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(commentRef01ListRt0), map[string]any{"id": commentRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func commentBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "comment", "CommentTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read comment test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse comment test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"comment01", "comment02", "comment03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("KOREANJSON_TEST_COMMENT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"KOREANJSON_TEST_COMMENT_ENTID": idmap,
		"KOREANJSON_TEST_LIVE":      "FALSE",
		"KOREANJSON_TEST_EXPLAIN":   "FALSE",
		"KOREANJSON_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["KOREANJSON_TEST_COMMENT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["KOREANJSON_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["KOREANJSON_APIKEY"],
			},
			extra,
		})
		client = sdk.NewKoreanJsonSDK(core.ToMapAny(mergedOpts))
	}

	live := env["KOREANJSON_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["KOREANJSON_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/comment_entity_test.lua
-- Comment entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("korean-json_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("CommentEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:Comment(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = comment_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "comment." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set KOREANJSON_TEST_COMMENT_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local comment_ref01_ent = client:Comment(nil)
    local comment_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.comment"), "comment_ref01"))

    local comment_ref01_data_result, err = comment_ref01_ent:create(comment_ref01_data, nil)
    assert.is_nil(err)
    comment_ref01_data = helpers.to_map(comment_ref01_data_result)
    assert.is_not_nil(comment_ref01_data)
    assert.is_not_nil(comment_ref01_data["id"])

    -- LIST
    local comment_ref01_match = {}

    local comment_ref01_list_result, err = comment_ref01_ent:list(comment_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(comment_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(comment_ref01_list_result),
      { id = comment_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local comment_ref01_data_up0_up = {
      id = comment_ref01_data["id"],
    }

    local comment_ref01_markdef_up0_name = "content"
    local comment_ref01_markdef_up0_value = "Mark01-comment_ref01_" .. tostring(setup.now)
    comment_ref01_data_up0_up[comment_ref01_markdef_up0_name] = comment_ref01_markdef_up0_value

    local comment_ref01_resdata_up0_result, err = comment_ref01_ent:update(comment_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local comment_ref01_resdata_up0 = helpers.to_map(comment_ref01_resdata_up0_result)
    assert.is_not_nil(comment_ref01_resdata_up0)
    assert.are.equal(comment_ref01_resdata_up0["id"], comment_ref01_data_up0_up["id"])
    assert.are.equal(comment_ref01_resdata_up0[comment_ref01_markdef_up0_name], comment_ref01_markdef_up0_value)

    -- LOAD
    local comment_ref01_match_dt0 = {
      id = comment_ref01_data["id"],
    }
    local comment_ref01_data_dt0_loaded, err = comment_ref01_ent:load(comment_ref01_match_dt0, nil)
    assert.is_nil(err)
    local comment_ref01_data_dt0_load_result = helpers.to_map(comment_ref01_data_dt0_loaded)
    assert.is_not_nil(comment_ref01_data_dt0_load_result)
    assert.are.equal(comment_ref01_data_dt0_load_result["id"], comment_ref01_data["id"])

    -- REMOVE
    local comment_ref01_match_rm0 = {
      id = comment_ref01_data["id"],
    }
    local _, err = comment_ref01_ent:remove(comment_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local comment_ref01_match_rt0 = {}

    local comment_ref01_list_rt0_result, err = comment_ref01_ent:list(comment_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(comment_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(comment_ref01_list_rt0_result),
      { id = comment_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function comment_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/comment/CommentTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read comment test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "comment01", "comment02", "comment03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("KOREANJSON_TEST_COMMENT_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["KOREANJSON_TEST_COMMENT_ENTID"] = idmap,
    ["KOREANJSON_TEST_LIVE"] = "FALSE",
    ["KOREANJSON_TEST_EXPLAIN"] = "FALSE",
  })

  local idmap_resolved = helpers.to_map(
    env["KOREANJSON_TEST_COMMENT_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["KOREANJSON_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["KOREANJSON_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["KOREANJSON_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
